### PR TITLE
Redefine Foundry test for Lon contract

### PR DESCRIPTION
**Pull request by oneleo**
To make the foundry test for Lon contract more readable, these test codes have been changed and added to the hierarchy.

These test codes conform to the following specifications:

1. One file, one contract.
2. Test contracts are named with upper camel case.
3. Test functions are named with lower camel case.
4. Minimal import principle: do not import unused libraries.
5. Uniformly put the Test Helper functions in the `setup.t.sol` file 

Please use the following command to test:
```
% forge test -vvv --match-path 'contracts/test/LON/*.t.sol' --force
```

The execution result is as follows:
```
Running 2 tests for contracts/test/LON/Burn.t.sol:TestLONBurn
[PASS] testBurn() (gas: 48457)
[PASS] testCannotBurnMoreThanOwned() (gas: 62928)
Test result: ok. 2 passed; 0 failed; finished in 10.96ms

Running 4 tests for contracts/test/LON/Mint.t.sol:TestLONMint
[PASS] testCannotMintByNotMinter() (gas: 13396)
[PASS] testCannotMintExceedCap() (gas: 15903)
[PASS] testCannotMintToZeroAddress() (gas: 10710)
[PASS] testMint() (gas: 59205)
Test result: ok. 4 passed; 0 failed; finished in 11.42ms

Running 1 test for contracts/test/LON/Constructor.t.sol:TestLONConstructor
[PASS] testConstructor() (gas: 15805)
Test result: ok. 1 passed; 0 failed; finished in 10.95ms

Running 2 tests for contracts/test/LON/SetMinter.t.sol:TestLONSetMinter
[PASS] testCannotSetByNotOwner() (gas: 13377)
[PASS] testSetMinter() (gas: 17421)
Test result: ok. 2 passed; 0 failed; finished in 2.49ms

Running 1 test for contracts/test/LON/EmergencyWithdraw.t.sol:TestLONEmergencyWithdraw
[PASS] testEmergencyWithdraw() (gas: 845700)
Test result: ok. 1 passed; 0 failed; finished in 10.50ms

Running 5 tests for contracts/test/LON/Permit.t.sol:TestLONPermit
[PASS] testCannotPermitByZeroAddress() (gas: 22613)
[PASS] testCannotPermitIfSignAgain() (gas: 80206)
[PASS] testCannotPermitWhenExpired() (gas: 22617)
[PASS] testCannotPermitWithInvalidSignature() (gas: 50750)
[PASS] testPermit() (gas: 81279)
Test result: ok. 5 passed; 0 failed; finished in 25.35ms
```

---

The old Foundry test for the Lon contract can also be tested with the following command:
```
% forge test -vvv --match-path 'contracts/test/LON.t.sol'

Running 15 tests for contracts/test/LON.t.sol:LONTest
[PASS] testBurn() (gas: 48476)
[PASS] testCannotBurnMoreThanOwned() (gas: 62906)
[PASS] testCannotMintBeyondCap() (gas: 15948)
[PASS] testCannotMintByNotMinter() (gas: 13418)
[PASS] testCannotMintToZeroAddress() (gas: 10733)
[PASS] testCannotPermitByZeroAddress() (gas: 22541)
[PASS] testCannotPermitWithExpiredPermit() (gas: 22678)
[PASS] testCannotPermitWithInvalidUserSig() (gas: 50744)
[PASS] testCannotPermitWithSameSignatureAgain() (gas: 80309)
[PASS] testCannotSetMinterByNotOwner() (gas: 13355)
[PASS] testEmergencyWithdraw() (gas: 845745)
[PASS] testMint() (gas: 59228)
[PASS] testPermit() (gas: 81273)
[PASS] testSetMinter() (gas: 17510)
[PASS] testSetupLON() (gas: 15827)
Test result: ok. 15 passed; 0 failed; finished in 12.81ms
```
Thank you very much!

### CODE REVIEW

File: contracts/test/LON/Burn.t.sol [hunk 1]
```diff
@@ -0,0 +1,25 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity 0.7.6;
+
+import "contracts-test/LON/Setup.t.sol";
+
+contract TestLONBurn is TestLON {
+    // include Snapshot struct from BalanceSnapshot
```
Comment (pilagod) [inline contracts/test/LON/Burn.t.sol:7]: I think this comment could be removed, also for other places.

File: contracts/test/LON/Burn.t.sol [hunk 2]
```diff
@@ -0,0 +1,25 @@
+// SPDX-License-Identifier: UNLICENSED
```
Comment (pilagod) [inline contracts/test/LON/Burn.t.sol:1]: Our repo is under MIT license, maybe our tests could also use MIT.

File: contracts/test/LON/Constructor.t.sol [hunk 3]
```diff
@@ -0,0 +1,13 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity 0.7.6;
+
+import "contracts-test/LON/Setup.t.sol";
+
+contract TestLONConstructor is TestLON {
+    // normal case
```
Comment (pilagod) [inline contracts/test/LON/Constructor.t.sol:7]: Could be removed

File: contracts/test/LON/Setup.t.sol [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "contracts/Lon.sol";
+import "contracts-test/utils/BalanceSnapshot.sol";
+
+contract TestLON is Test {
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+    using BalanceSnapshot for BalanceSnapshot.Snapshot;
```
Comment (pilagod) [inline contracts/test/LON/Setup.t.sol:14]: It seems that these three libs are not used in setup contract.

File: contracts/test/LON/Setup.t.sol [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    uint256 DEADLINE = block.timestamp + 1;
+
+    struct Permit {
+        address owner;
+        address spender;
+        uint256 value;
+        uint256 nonces;
+        uint256 deadline;
+    }
+    Permit DEFAULT_PERMIT;
```
Comment (pilagod) [inline contracts/test/LON/Setup.t.sol:34]: Permit seems to be used only in `permit` test. Maybe code related to `permit` could be collected in permit test contract?

File: contracts/test/LON/Burn.t.sol (hunk 1 above)
Comment (oneleo) [reply contracts/test/LON/Burn.t.sol]: Sure! I will remove these no longer used comments in the next commit.

File: contracts/test/LON/Burn.t.sol (hunk 2 above)
Comment (oneleo) [reply contracts/test/LON/Burn.t.sol]: Certainly! I will change these code license to MIT license in the next commit.

File: contracts/test/LON/Constructor.t.sol (hunk 3 above)
Comment (oneleo) [reply contracts/test/LON/Constructor.t.sol]: OK!

File: contracts/test/LON/Setup.t.sol (hunk 4 above)
Comment (oneleo) [reply contracts/test/LON/Setup.t.sol]: Yes!

File: contracts/test/LON/Setup.t.sol (hunk 5 above)
Comment (oneleo) [reply contracts/test/LON/Setup.t.sol]: No problem, I will use the setUp() inheritance way to downsize Setup.t.sol and move permit struct to Permit.t.sol file in the next commit

File: contracts/test/LON/Permit.t.sol [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+        (uint8 v, bytes32 r, bytes32 s) = vm.sign(privateKey, EIP712SignDigest);
+        return (v, r, s);
+    }
+
+    function _getEIP712Hash(bytes32 structHash) internal view returns (bytes32) {
+        string memory EIP191_HEADER = "\x19\x01";
+        bytes32 DOMAIN_SEPARATOR = lon.DOMAIN_SEPARATOR();
+        return keccak256(abi.encodePacked(EIP191_HEADER, DOMAIN_SEPARATOR, structHash));
+    }
```
Comment (pilagod) [inline contracts/test/LON/Permit.t.sol:109]: I feel this is a general util not only for permit. Maybe this is more suitable to place in setup.

Comment (oneleo) [reply contracts/test/LON/Permit.t.sol]: Yes! I also feel that way. I've Moved the _getEIP712Hash general util function from the Permit.t.sol test file to Setup.t.sol test contract.

File: contracts/test/LON/Burn.t.sol [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+        vm.expectRevert("ERC20: burn amount exceeds balance");
+        vm.prank(user);
+        lon.burn(excessAmount);
+    }
+
+    function testBurn() public {
+        lon.mint(user, 1e18);
+        BalanceSnapshot.Snapshot memory userLon = BalanceSnapshot.take(user, address(lon));
+        vm.prank(user);
+        lon.burn(1e18);
```
Comment (charlesjhongc) [inline contracts/test/LON/Burn.t.sol:21]: Minor suggestion : declare a `uint` to replace `1e18`.

**Review by charlesjhongc — DISMISSED** [review]
LGTM.

**Review by NIC619 — DISMISSED** [review]

File: contracts/test/LON/Burn.t.sol (hunk 7 above)
Comment (oneleo) [reply contracts/test/LON/Burn.t.sol:21]: No problem! I think this's better! I have take the exact declared variable as the parameters of the functions.

**Review by charlesjhongc — APPROVED** [review]
LGTM

**Review by pilagod — APPROVED** [review]

**Review by NIC619 — APPROVED** [review]

**Pull request merged**